File: frontend/src/utils/evalStatus.js
// Lifecycle status of a single eval result. Mirrors the backend
// `EvalEntryStatus` (tracer EvalLogger.status): pending -> running ->
// completed | errored | skipped. The read APIs surface the non-score states
// (pending/running/skipped) so the UI can render a loading / queued / skipped
// cell instead of a misleading blank or 0.
export const EVAL_STATUS = {
  PENDING: "pending",
  RUNNING: "running",
  COMPLETED: "completed",
  ERRORED: "errored",
  SKIPPED: "skipped",
};

const NON_SCORE_LABELS = {
  [EVAL_STATUS.PENDING]: "Queued",
  [EVAL_STATUS.RUNNING]: "Evaluating…",
  [EVAL_STATUS.SKIPPED]: "Skipped",
};

// Normalize a backend status string to one of the non-score lifecycle states
// the UI renders specially, or null when it is a real/terminal result.
export const getEvalNonScoreStatus = (status) => {
  const s = String(status || "").toLowerCase();
  return s in NON_SCORE_LABELS ? s : null;
};

// Pull a non-score lifecycle marker out of a backend eval cell value, which may
// be a scalar score, an `{ error: true }` marker, or a `{ status: ... }` marker.
export const getEvalNonScoreStatusFromValue = (value) => {
  if (!value || typeof value !== "object" || Array.isArray(value)) return null;
  return getEvalNonScoreStatus(value.status || value.eval_status);
};

export const getEvalStatusLabel = (status) =>
  NON_SCORE_LABELS[getEvalNonScoreStatus(status)] || "";

File: futureagi/tracer/services/clickhouse/query_builders/eval_status.py
"""Shared resolution of a non-terminal / skipped eval cell state.

The span / trace / voice-call eval pivots share one precedence for a
``(row, eval_config)`` pair that has *no completed result and no error*:
``skipped > running > pending``. The completed (score) and errored states are
resolved by each caller (they own the score/error shapes); this helper only
decides the lifecycle marker when neither applies, from the per-status counts
the eval query selects.
"""

from __future__ import annotations

from typing import Any


def non_terminal_eval_marker(row: dict[str, Any]) -> dict[str, Any] | None:
    """Cell marker for a (row, config) pair with no completed score and no error.

    Returns ``{"status": "skipped"|"running"|"pending"}`` (with ``skipped_reason``
    when present) per the precedence ``skipped > running > pending``, or ``None``
    when no eval has run for the pair (renders blank — "no eval run").
    """
    if (row.get("skipped_count", 0) or 0) > 0:
        marker: dict[str, Any] = {"status": "skipped"}
        reason = row.get("skipped_reason")
        if reason:
            marker["skipped_reason"] = reason
        return marker
    if (row.get("running_count", 0) or 0) > 0:
        return {"status": "running"}
    if (row.get("pending_count", 0) or 0) > 0:
        return {"status": "pending"}
    return None
